### PR TITLE
Improve excludes

### PR DESCRIPTION
**Pull request by dkrutskikh**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix
[ ] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### What changes did you make? (Give an overview)
Closie #68

### CODE REVIEW

File: lib/src/analysis_options.dart [hunk 1]
```diff
@@ -25,74 +32,85 @@ class AnalysisOptions {
...
-    final node = loadYamlNode(content ?? '');
-    if (node is YamlMap && node.nodes[_rootKey] is YamlMap) {
-      final metricsOptions = node.nodes[_rootKey] as YamlMap;
-
-      if (_isYamlMapOfStringsAndIntegers(metricsOptions.nodes[_metricsKey])) {
-        final configMap = metricsOptions[_metricsKey] as YamlMap;
+    final metricsOptions = configMap[_rootKey];
+    if (metricsOptions != null && metricsOptions is Map<String, Object>) {
+      final configMap = metricsOptions[_metricsKey];
+      if (configMap != null && configMap is Map<String, Object>) {
```
Comment (incendial) [inline lib/src/analysis_options.dart:53]: May be move this check to separate function like `isMap` for all same usages on lines 51, 53, 83, 100?

File: lib/src/analysis_options.dart [hunk 2]
```diff
@@ -25,74 +32,85 @@ class AnalysisOptions {
...
+          numberOfMethodsWarningLevel: configMap['number-of-methods'].as<int>(),
         );
       }
 
-      if (_isYamlListOfStrings(metricsOptions.nodes[_metricsExcludeKey])) {
-        metricsExcludePatterns =
-            List.unmodifiable(metricsOptions[_metricsExcludeKey] as Iterable);
+      final excludeList = metricsOptions[_metricsExcludeKey];
+      if (excludeList != null &&
+          excludeList is Iterable<Object> &&
```
Comment (incendial) [inline lib/src/analysis_options.dart:66]: Same suggestion as for map, but for iterable.

File: lib/src/analysis_options.dart [hunk 3]
```diff
@@ -25,74 +32,85 @@ class AnalysisOptions {
...
+            return node != null && node is Map<String, Object>;
+          }).map((key) => MapEntry(key, rulesNode[key] as Map<String, Object>)),
         ]));
       }
     }
 
+    final analyzerOptions = configMap[_analyzerKey];
+    if (analyzerOptions != null && analyzerOptions is Map<String, Object>) {
+      final excludeList = analyzerOptions[_excludeKey];
+      if (excludeList != null && excludeList is List<Object>) {
```
Comment (incendial) [inline lib/src/analysis_options.dart:102]: Can it be `excludeList is Iterable<Object>` instead of a `List`?

File: lib/src/utils/yaml_utls.dart [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import 'package:yaml/yaml.dart';
```
Comment (Anrock) [inline lib/src/utils/yaml_utls.dart:4]: Typo in file name. yaml_utls.dart => yaml_ut**i**ls.dart

File: lib/src/analysis_options.dart [hunk 5]
```diff
@@ -29,63 +30,63 @@ class AnalysisOptions {
...
+    Iterable<String> metricsExcludePatterns = <String>[];
     var rules = <String, Map<String, Object>>{};
 
-    if (node.nodes[_rootKey] is YamlMap) {
-      final metricsOptions = node.nodes[_rootKey] as YamlMap;
-
-      if (_isYamlMapOfStringsAndIntegers(metricsOptions.nodes[_metricsKey])) {
-        final configMap = metricsOptions[_metricsKey] as YamlMap;
+    final metricsOptions = configMap[_rootKey];
+    if (metricsOptions != null && metricsOptions is Map<String, Object>) {
```
Comment (Anrock) [inline lib/src/analysis_options.dart:51]: `null is T` returns false (which is kinda strange given that null is member of everytype), so there is no need for dedicated null-check.

File: lib/src/utils/yaml_utls.dart [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+/// Merge lists, avoiding duplicates.
+List<Object> _mergeLists(List<Object> defaults, List<Object> overrides) =>
+    List.unmodifiable(<Object>{...defaults, ...overrides});
+
+bool _isListOfStrings(Object object) =>
+    object != null &&
+    object is List<Object> &&
+    object.every((node) => node is String);
+
+Map<String, bool> _listToMap(List<Object> list) => list != null
```
Comment (incendial) [inline lib/src/utils/yaml_utls.dart:108]: Do you really need to check for null here? In all usages you first invoke` _isListOfStrings` which is also checking for null.

**Comment by incendial** [comment]
MR is not linked to issue properly (you have a typo).

File: lib/src/utils/yaml_utls.dart (hunk 6 above)
Comment (dkrutskikh) [reply lib/src/utils/yaml_utls.dart]: fixed

File: lib/src/analysis_options.dart (hunk 5 above)
Comment (dkrutskikh) [reply lib/src/analysis_options.dart]: fixed

File: lib/src/utils/yaml_utls.dart (hunk 4 above)
Comment (dkrutskikh) [reply lib/src/utils/yaml_utls.dart]: fixed

File: lib/src/analysis_options.dart (hunk 3 above)
Comment (dkrutskikh) [reply lib/src/analysis_options.dart]: fixed

File: lib/src/analysis_options.dart (hunk 2 above)
Comment (dkrutskikh) [reply lib/src/analysis_options.dart]: fixed

File: lib/src/analysis_options.dart (hunk 1 above)
Comment (dkrutskikh) [reply lib/src/analysis_options.dart]: fixed

**Pull request merged**